### PR TITLE
Proposal add psp rules support

### PR DESCRIPTION
**Pull request by mstemm**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

/kind documentation

> /kind failing-test

> /kind feature

> /kind flaky-test

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area engine

> /area rules

> /area deployment

> /area integrations

> /area examples

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: proposals/20190909-psp-rules-support.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+## Motivation
+
+PSPs provide a rich powerful framework to restrict the behavior of pods and apply consistent security policies across a cluster, but it’s difficult to know the gap between what you want your security policy to be and what your cluster is actually doing. Additionally, since PSPs enforce once applied, they might prevent pods from running, and the process of tuning a PSP live on a cluster can be disruptive and painful.
+
+That's where Falco comes in. We want to make it possible for Falco to perform a “dry run” evaluation of a PSP, translating it to Falco rules that observe the behavior of deployed pods and sending alerts for violations, *without* blocking. This helps accelerate the authoring cycle, providing a complete authoring framework for PSPs without deploying straight to the cluster.
+
+### Goals
+
+Transparently read a candidate PSP into an equivalent set of falco rules that can look for the conditions in the PSP.
```
Comment (kris-nova) [inline proposals/20190909-psp-rules-support.md]: Can we have more technical detail here? Are we generating Falco rules as a file? as a pipe? as shared memory?

File: proposals/20190909-psp-rules-support.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+### Non-Goals
+
+Falco will not automatically read PSPs from a cluster, will not install PSPs, and will not provide guidance on the parts of your infrastructure that are already covered by PSPs. This feature only helps with the testing part of a candidate PSP. For coming up with an initial PSP, you can use tools like [https://github.com/sysdiglabs/kube-psp-advisor](Kube PSP Advisor).
+
+## Proposal
+
+### Use cases
+
+You'll be able to run falco with a `--psp` argument that provides a single PSP Yaml file. Falco will automatcially convert the PSP into an equivalent set of falco rules, load the rules, and then run with the loaded rules. You can optionally provide a `--psp_save=<path>` command line option to save the converted rules to a file.
```
Comment (kris-nova) [inline proposals/20190909-psp-rules-support.md]: Arguments to the CLI are effectively an `API` 

My only suggestion here would be that we start using some sort of prefix to denote this is Kubernetes specific.

For example: 

```
--k8s-pod-security-policy
--k8s-psp
--enable-kubernetes-pod-security-policy
```

File: proposals/20190909-psp-rules-support.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+### Diagrams
+
+No diagrams yet.
+
+### Design Details
+
+* We'll use [inja](https://github.com/pantor/inja) as the templating engine.
```
Comment (kris-nova) [inline proposals/20190909-psp-rules-support.md]: I think this eludes to my question above about what medium we use to create rules. 

In general I think that templating anything is a sign that we need an input `API` with clear methods. In my mind being able to provide tangible input dynamically at runtime seems to be the most effective. 

---

Also what's the dependency plan here?

We are introducing a new dependency into the compiling process, do we want to document this?

File: proposals/20190909-psp-rules-support.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+That's where Falco comes in. We want to make it possible for Falco to perform a “dry run” evaluation of a PSP, translating it to Falco rules that observe the behavior of deployed pods and sending alerts for violations, *without* blocking. This helps accelerate the authoring cycle, providing a complete authoring framework for PSPs without deploying straight to the cluster.
+
+### Goals
+
+Transparently read a candidate PSP into an equivalent set of falco rules that can look for the conditions in the PSP.
+
+### Non-Goals
+
+Falco will not automatically read PSPs from a cluster, will not install PSPs, and will not provide guidance on the parts of your infrastructure that are already covered by PSPs. This feature only helps with the testing part of a candidate PSP. For coming up with an initial PSP, you can use tools like [https://github.com/sysdiglabs/kube-psp-advisor](Kube PSP Advisor).
```
Comment (kris-nova) [inline proposals/20190909-psp-rules-support.md]: I think most folks in Kubernetes expect dynamic compliance. If a user defines a `PSP` in a cluster, the expectation is that Kubernetes will reconcile and be eventually consistent and enforce whatever is defined.

File: proposals/20190909-psp-rules-support.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+- [Proposal](#proposal)
+  * [Use cases](#use-cases)
+  * [Diagrams](#diagrams)
+  * [Design Details](#design-details)
+
+<!-- tocstop -->
+
+## Summary
+
+We want to make it easier for K8s Cluster Operators to Author Pod Security Policies by providing a way to read a PSP, convert it to a set of falco rules, and then run falco with those rules.
```
Comment (kris-nova) [inline proposals/20190909-psp-rules-support.md]: Yes! This would be amazing for the project! Strong +1 from me

Comment (fntlnz) [reply proposals/20190909-psp-rules-support.md]: Where is the convertion done? In a cli ? In falco itself?

File: proposals/20190909-psp-rules-support.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+## Motivation
+
+PSPs provide a rich powerful framework to restrict the behavior of pods and apply consistent security policies across a cluster, but it’s difficult to know the gap between what you want your security policy to be and what your cluster is actually doing. Additionally, since PSPs enforce once applied, they might prevent pods from running, and the process of tuning a PSP live on a cluster can be disruptive and painful.
```
Comment (fntlnz) [inline proposals/20190909-psp-rules-support.md]: How we guarantee that our mapping is always 1:1 between what kubernetes does and what falco does?

File: proposals/20190909-psp-rules-support.md (hunk 2 above)
Comment (fntlnz) [reply proposals/20190909-psp-rules-support.md]: I don't like the falco daemon cli to also do cli related tasks, this should be done in falcoctl in my opinion

File: proposals/20190909-psp-rules-support.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+## Proposal
+
+### Use cases
+
+You'll be able to run falco with a `--psp` argument that provides a single PSP Yaml file. Falco will automatcially convert the PSP into an equivalent set of falco rules, load the rules, and then run with the loaded rules. You can optionally provide a `--psp_save=<path>` command line option to save the converted rules to a file.
+
+### Diagrams
+
+No diagrams yet.
```
Comment (fntlnz) [inline proposals/20190909-psp-rules-support.md]: It will be helpful

File: proposals/20190909-psp-rules-support.md (hunk 3 above)
Comment (fntlnz) [reply proposals/20190909-psp-rules-support.md]: I'm very scared about this, we need to have a maintainers discussion on wether we want yet another parsing engine in Falco.

File: proposals/20190909-psp-rules-support.md (hunk 1 above)
Comment (mstemm) [reply proposals/20190909-psp-rules-support.md]: The PSP is converted into a set of falco rules which can be either saved as a file for later use/inspection, or loaded directly so they they can monitor system calls and k8s audit activity.

File: proposals/20190909-psp-rules-support.md (hunk 2 above)
Comment (mstemm) [reply proposals/20190909-psp-rules-support.md]: I prefer having a single program handle both the steps of conversion and loading the generated rules, as I feel that the fewer steps involved, the easier it is to get set up and running.

File: proposals/20190909-psp-rules-support.md (hunk 3 above)
Comment (mstemm) [reply proposals/20190909-psp-rules-support.md]: I'm not worried about adding new third-party code, in the last ~9 months we've added njson, tbb, civetweb, and obviously protobuf/grpc is being added for grpc support. I imagine we'll add prometheus soon so we can collect and export prometheus metrics.

inja isn't really used to parse or interpret anything. It's used to specify a skeleton of what the rules that validate the conditions in a PSP would look like, which gets combined with options/lists/etc to result in a set of rules that validate the conditions in the PSP.

Here's an example of what the template looks like: https://github.com/falcosecurity/falco/blob/psp-rules-support/rules/templates/k8s_psp_rules.yaml.tmpl. I'm not really sure how you can put an API around that, but I'm open to feedback.

File: proposals/20190909-psp-rules-support.md (hunk 4 above)
Comment (mstemm) [reply proposals/20190909-psp-rules-support.md]: The use case here is for cluster operators who want to author PSPs, but don't want to just put it in a cluster and see what breaks. For example, if your PSP sets `privileged` to false, but it turns out some of your pods are running privileged, they won't be able to start.

With this feature, they could iterate without enforcement until they have a PSP that matches the actual behavior of their cluster. Some of that will come from changing the PSP, some of that will come from changng the behavior of the cluster. The important part is that it's not mistakenly preventing things from running while you're figuring it out.

File: proposals/20190909-psp-rules-support.md (hunk 6 above)
Comment (mstemm) [reply proposals/20190909-psp-rules-support.md]: It's a risk, but I tried really hard to be comprehensive and added a big set of tests to verify all the parts of the generated rules match the activity covered by the PSP. You can see the tests I wrote at https://github.com/falcosecurity/falco/blob/psp-rules-support/test/falco_tests_psp.yaml.

File: proposals/20190909-psp-rules-support.md (hunk 5 above)
Comment (mstemm) [reply proposals/20190909-psp-rules-support.md]: It's in falco. Happy to discuss whether the conversion occurs in falco or outside--I lean towards in falco as it involves fewer moving parts, but obviously there are other opinions.

Remember, though, that the bulk of the work is in implementing the rules that result from the PSP, which requires new operators and fields to work on audit logs.

File: proposals/20190909-psp-rules-support.md (hunk 6 above)
Comment (fntlnz) [reply proposals/20190909-psp-rules-support.md]: OK thanks for clarifying

File: proposals/20190909-psp-rules-support.md (hunk 3 above)
Comment (fntlnz) [reply proposals/20190909-psp-rules-support.md]: Ok got your point, thanks for clarifiying

Comment (fntlnz) [reply proposals/20190909-psp-rules-support.md]: to clarify: grpc was already there for CRI, it wasn't added for grpc support

File: proposals/20190909-psp-rules-support.md (hunk 5 above)
Comment (leodido) [inline proposals/20190909-psp-rules-support.md]:
```suggestion
We want to make it easier for K8s Cluster Operators to Author Pod Security Policies by providing a way to read a PSP, convert it to a set of Falco rules, and then run Falco with those rules.
```

File: proposals/20190909-psp-rules-support.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+## Summary
+
+We want to make it easier for K8s Cluster Operators to Author Pod Security Policies by providing a way to read a PSP, convert it to a set of falco rules, and then run falco with those rules.
+
+## Motivation
+
+PSPs provide a rich powerful framework to restrict the behavior of pods and apply consistent security policies across a cluster, but it’s difficult to know the gap between what you want your security policy to be and what your cluster is actually doing. Additionally, since PSPs enforce once applied, they might prevent pods from running, and the process of tuning a PSP live on a cluster can be disruptive and painful.
+
+That's where Falco comes in. We want to make it possible for Falco to perform a “dry run” evaluation of a PSP, translating it to Falco rules that observe the behavior of deployed pods and sending alerts for violations, *without* blocking. This helps accelerate the authoring cycle, providing a complete authoring framework for PSPs without deploying straight to the cluster.
```
Comment (leodido) [inline proposals/20190909-psp-rules-support.md]:
```suggestion
That's where Falco comes in. We want to make it possible for Falco to perform a "dry run" evaluation of a PSP, translating it to Falco rules that observe the behaviour of deployed pods and sending alerts for violations, *without* blocking. This helps accelerate the authoring cycle, providing a complete authoring framework for PSPs without deploying straight to the cluster.
```

File: proposals/20190909-psp-rules-support.md (hunk 1 above)
Comment (leodido) [inline proposals/20190909-psp-rules-support.md]:
```suggestion
Transparently read a candidate PSP into an equivalent set of Falco rules that can look for the conditions in the PSP.

The PSP is converted into a set of Falco rules which can be either saved as a file for later use/inspection, or loaded directly so they they can monitor system calls and k8s audit activity.
```

File: proposals/20190909-psp-rules-support.md (hunk 4 above)
Comment (leodido) [inline proposals/20190909-psp-rules-support.md]:
```suggestion
Falco will not automatically read PSPs from a cluster, will not install PSPs, and will not provide guidance on the parts of your infrastructure that are already covered by PSPs. This feature only helps with the testing part of a candidate PSP. For coming up with an initial PSP, you can use tools like [https://github.com/sysdiglabs/kube-psp-advisor](Kube PSP Advisor).

The use case here is for cluster operators who want to author PSPs, but don't want to just put it in a cluster and see what breaks. For example, if your PSP sets privileged to false, but it turns out some of your pods are running privileged, they won't be able to start.

With this feature, they could iterate without enforcement until they have a PSP that matches the actual behaviour of their cluster. Some of that will come from changing the PSP, some of that will come from changing the behaviour of the cluster. The important part is that it's not mistakenly preventing things from running while you're figuring it out.
```

File: proposals/20190909-psp-rules-support.md (hunk 2 above)
Comment (leodido) [inline proposals/20190909-psp-rules-support.md]:
```suggestion
You'll be able to run falco with a `--psp` argument that provides a single PSP yaml file. Falco will automatically convert the PSP into an equivalent set of Falco rules, load the rules, and then run with the loaded rules. You can optionally provide a `--psp_save=<path>` command line option to save the converted rules to a file.
```

**Pull request merged**